src/main/python/run_trecqa.py

```diff
@@ -1,223 +1,297 @@
+                label = label.group(1)
+                label = label == "positive"
+                answer = line.lower().split("\t")
+                answer_count += 1
 
-        if qid not in labels:
-          labels[qid] = {}
+                answer_id = "Q{}-A{}".format(qid, answer_count)
+                answers[qid].append((answer_id, answer, label))
+                labels[qid][answer_id] = label
+            prev = line
 
-      if prev and prev.startswith('<question>'):
-        questions.append(line)
+    return questions, answers, labels
 
-      label = re.match('^<(positive|negative)>', prev)
 
-      if label:
-        label = label.group(1)
-        label = 1 if label == 'positive' else 0
-        answer = line.split('\t')
-        answer_count += 1
+def create_qrel_jaccard(fname, answers):
+    with open(fname, "w") as f:
+        for qid in answers.keys():
+            for answer in answers[qid]:
+                f.write("{} 0 {} {}\n".format(qid, answer[0], answer[2]))
 
-        answer_id = 'Q{}-A{}'.format(qid, answer_count)
-        answers[qid].append((answer_id, answer, label))
-        labels[qid][answer_id] = label
-      prev = line
+# evaluate by TREC QA patterns
+def eval_by_pattern(qid, candidates, pattern, out):
+    seen = set([])
 
-  return questions, answers, labels
+    for i, cand in enumerate(candidates):
+        this_candidate = " ".join(cand[0])
 
+        hash_value = md5(this_candidate.encode()).hexdigest()
+        answer_id = "QA{}.{}".format(qid, hash_value)
 
-def create_ground_truth(fname, answers):
-  with open(fname, 'w') as f:
-    for qid in answers.keys():
-      for answer in answers[qid]:
-        f.write('{} 0 {} {}\n'.format(qid, answer[0], answer[2]))
+        if answer_id in seen:
+            continue
+        seen.add(answer_id)
 
+        out.write("{} Q0 {} {} {} TrecQA-pattern\n".format(qid, answer_id, i+1, 1.0/(i + 1)))
 
-def get_precision(actual, predicted, k):
-  total_count = 0.0
-  hits = 0.0
+# evaluate the run file at different depths
+def evaluate_at_k(fname, eval_depth):
+    for k in eval_depth:
+        with open(fname) as f, open("{}.k{}.txt".format(fname, k), "w") as output_file:
+            for line in f:
+                det = line.strip().split()
+                rank = int(det[3])
 
-  for qid in predicted.keys():
-    run_labels = predicted.copy()
+                if rank <= int(k):
+                    output_file.write(line)
 
-    if len(predicted[qid]) > k:
-      run_labels = run_labels[qid][:k]
-    else:
-      run_labels = run_labels[qid]
 
-    for p in run_labels:
-      total_count += 1
+def create_qrel_pattern(all_sentences, pattern, qrels, qid):
+    seen = set([])
+    for i, cand in enumerate(all_sentences):
+        this_candidate = cand.lower().split("\t")[0]
+        result = re.findall(r"{}".format(pattern.lower()), this_candidate)
 
-      if not actual[qid]:
-        continue
+        # use the hash of the candidates as the docid
```

Comment:
_Why_ are we using the hash of the candidate? Is it perhaps because the sentences don't _have_ docids?